src/filters/BallFilter.cpp

```diff
@@ -0,0 +1,139 @@
+
+BallFilter::BallFilter(const proto::SSL_DetectionBall &detectionBall, double detectTime, int cameraID) :
+        CameraFilter(detectTime, cameraID),
+        lastPredictTime{detectTime} {
+    KalmanInit(detectionBall);
+}
+void BallFilter::KalmanInit(const proto::SSL_DetectionBall &detectionBall) {
+    // SSL units are in mm, we do everything in SI units.
+    double x = mmToM(detectionBall.x());//m
+    double y = mmToM(detectionBall.y());//m
+    Kalman::Vector startState = {
+            x, y, 0, 0
+    };
+
+    Kalman::Matrix startCov;
+    startCov.eye();
+    //initial noise estimates
+    const double startPosNoise = 0.05;
+    startCov.at(0, 0) = startPosNoise;//m noise in x
+    startCov.at(1, 1) = startPosNoise;//m noise in y
+
+    kalman = std::make_unique<Kalman>(startState, startCov);
+
+    kalman->H.eye();     // Our observations are simply what we see.
+
+
+}
+void BallFilter::applyObservation(const proto::SSL_DetectionBall &detectionBall, int cameraID) {
+    Kalman::VectorO observation = {mmToM(detectionBall.x()), mmToM(detectionBall.y())};
+    //TODO: do things with the other ball fields (pixel pos, area)
+    kalman->z = observation;
+
+    //Observations which are not from the main camera are added but are seen as much more noisy
+    const double posVar = 0.02; //variance TODO: tune these 2
+    const double posVarOtherCamera = 0.05;
+    kalman->R.zeros();
+    if (cameraID == mainCamera) {
+        kalman->R.at(0, 0) = posVar;
+        kalman->R.at(1, 1) = posVar;
+    } else {
+        kalman->R.at(0, 0) = posVarOtherCamera;
+        kalman->R.at(1, 1) = posVarOtherCamera;
+    }
+    kalman->update();
+}
+proto::WorldBall BallFilter::asWorldBall() const {
+    proto::WorldBall msg;
+    const Kalman::Vector &state = kalman->state();
+    msg.mutable_pos()->set_x(state[0]);
+    msg.mutable_pos()->set_y(state[1]);
+    msg.mutable_vel()->set_x(state[2]);
+    msg.mutable_vel()->set_y(state[3]);
+    msg.set_visible(ballIsVisible());
+    //TODO: add height filter here, and actually set the z, z_vel, area fields
+    return msg;
+}
+double BallFilter::distanceTo(double x, double y) const {
+    const Kalman::Vector &state = kalman->state();
+    double dx = state[0] - mmToM(x);
+    double dy = state[1] - mmToM(y);
+    return sqrt(dx * dx + dy * dy);
+}
+void BallFilter::predict(double time, bool permanentUpdate, bool cameraSwitched) {
+    double dt = time - lastUpdateTime;
+    // forward model:
+    kalman->F.eye();
+    kalman->F.at(0, 2) = dt;
+    kalman->F.at(1, 3) = dt;
+
+    //TODO: add 2 stage forward model?
+    //Set B
+    kalman->B = kalman->F;
+    //Set u (we have no control input at the moment)
+    kalman->u.zeros();
+
+    //Set Q matrix
+    const float posNoise = 0.1;//TODO: tune
+    Kalman::MatrixO G;
+    G.zeros();
+    G.at(0, 0) = dt * posNoise;
+    G.at(0, 2) = 1 * posNoise;
+    G.at(1, 1) = dt * posNoise;
+    G.at(1, 3) = 1 * posNoise;
+    if (cameraSwitched){
```

Comment:
Is this the additional noise added to the camera that is not preferred so you can artificially weigh the preferred one more?

Reply:
This is not called the camera is not preferred, this is only called when you switch the main camera to another camera to avoid the middle-line effect. If you are receiving data from 2 camera's simultaneously, the main camera is more trusted than the other but this is done in the measurement noise and not in the process noise.

Reply:
What I meant in the comments on this part is that I was not sure how the non-main camera was being weighed less, and also how you decide which camera should be weighed less

Reply:
I understand your question, but that part is not even happening in this function (predict() ). Predict() does not change the camera; only update() can, since then is when we receive a camera frame. This code is only called for the one time the kalman filter performs a prediction after the main camera switched from one ID to another ID. The other camera being 'weighed' less is done by giving it higher measurement noise in applyObservation